x-pack/plugin/autoscaling/src/test/java/org/elasticsearch/xpack/autoscaling/storage/ReactiveStorageDeciderDecisionTests.java

```diff
@@ -218,7 +218,11 @@ public void testStoragePreventsMove() {
         do {
             startRandomShards();
             // all of the relevant replicas are assigned too.
-        } while (state.getRoutingNodes().unassigned().stream().map(ShardRouting::shardId).anyMatch(warmShards::contains));
+        } while (haveNonEmptyIntersection(shardIds(state.getRoutingNodes().unassigned()), warmShards)
+            || haveNonEmptyIntersection(
+                shardIds(RoutingNodesHelper.shardsWithState(state.getRoutingNodes(), ShardRoutingState.INITIALIZING)),
+                warmShards
+            ));
```

Comment:
Following code moves STARTED `warmShards` to the warm nodes (as it is impossible to move INITIALIZING shards), however some of the `warmShards` remained initializing after this loop due to ongoing rebalancing.

For the seed suggested in https://github.com/elastic/elasticsearch/issues/93817 we pick `{[test-7][0], [test-7][1], [test-7][3]}`, however `[test-7][3]` is not initialized.
```
routing_nodes:
-----node_id[hot_1][V]
--------[test-7][4], node[hot_1], [R], recovery_source[peer recovery], s[INITIALIZING], a[id=jCDltr6LSSq3QeoNg0APXA], unassigned_info[[reason=INDEX_CREATED], at[2023-04-04T14:45:02.794Z], delayed=false, allocation_status[no_attempt]], failed_attempts[0], expected_shard_size[1]
-----node_id[hot_0][V]
--------[test-7][0], node[hot_0], [R], recovery_source[peer recovery], s[INITIALIZING], a[id=2rUsaPQdS5SWys9z-H6fng], unassigned_info[[reason=INDEX_CREATED], at[2023-04-04T14:45:02.794Z], delayed=false, allocation_status[no_attempt]], failed_attempts[0], expected_shard_size[1]
--------[test-7][3], node[hot_0], relocating [hot_2], [P], recovery_source[peer recovery], s[INITIALIZING], a[id=fe1k-AAKSPmKJ3g6mcAXCw, rId=A41h2dm5T-WtdRO1bj-2SA], failed_attempts[0], expected_shard_size[0]
        ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
-----node_id[hot_3][V]
--------[test-7][1], node[hot_3], [R], recovery_source[peer recovery], s[INITIALIZING], a[id=uttrNnlPSHu_vwR58LJ2BA], unassigned_info[[reason=INDEX_CREATED], at[2023-04-04T14:45:02.794Z], delayed=false, allocation_status[no_attempt]], failed_attempts[0], expected_shard_size[1]
--------[test-7][2], node[hot_3], [P], s[STARTED], a[id=GDO3KyBHRd-qWFVrOxYJRQ], failed_attempts[0]
-----node_id[hot_2][V]
--------[test-7][2], node[hot_2], [R], recovery_source[peer recovery], s[INITIALIZING], a[id=_BsL8zmbRFGK-64C3vp9Og], unassigned_info[[reason=INDEX_CREATED], at[2023-04-04T14:45:02.794Z], delayed=false, allocation_status[no_attempt]], failed_attempts[0], expected_shard_size[1]
--------[test-7][3], node[hot_2], relocating [hot_0], [P], s[RELOCATING], a[id=A41h2dm5T-WtdRO1bj-2SA, rId=fe1k-AAKSPmKJ3g6mcAXCw], failed_attempts[0], expected_shard_size[0]
-----node_id[hot_4][V]
--------[test-7][3], node[hot_4], [R], recovery_source[peer recovery], s[INITIALIZING], a[id=IG5OmNBRRy68DhIlTbpI4Q], unassigned_info[[reason=INDEX_CREATED], at[2023-04-04T14:45:02.794Z], delayed=false, allocation_status[no_attempt]], failed_attempts[0], expected_shard_size[1]
--------[test-7][4], node[hot_4], [P], s[STARTED], a[id=wo_Bx2xqTiyXi3rT09qa6g], failed_attempts[0]

-----node_id[warm_1][V]
--------[test-7][0], node[warm_1], [P], s[STARTED], a[id=XdQDlbUnRHutgARxzBYiMQ], failed_attempts[0]
-----node_id[warm_0][V]
--------[test-7][1], node[warm_0], [P], s[STARTED], a[id=5dlcuq_LQ42gIOYEfcojHw], failed_attempts[0]

---- unassigned
```